test/test_linalg.py

```diff
@@ -4758,6 +4758,102 @@ def check(*args, regex, exception=RuntimeError):
         check(x, [-1], regex=r'not within the valid range \[0, 52\)', exception=ValueError)
         check(x, [52], regex=r'not within the valid range \[0, 52\)', exception=ValueError)
 
+    def _gen_shape_inputs_linalg_triangular_solve(self, shape, dtype, device, well_conditioned=False):
+        make_arg = partial(make_tensor, dtype=dtype, device=device)
+        b, n, k = shape
+        for left, uni, expand_a, tr_a, conj_a, expand_b, tr_b, conj_b in product((True, False), repeat=8):
+            if (conj_a or conj_b) and not dtype.is_complex:
+                continue
+            # We just expand on the batch size
+            if (expand_a or expand_b) and b == 1:
+                continue
+
+            size_a = (b, n, n) if left else (b, k, k)
+            size_b = (b, n, k) if not tr_b else (b, k, n)
+
+            # If expand_a or expand_b, we'll expand them to the correct size later
+            if b == 1 or expand_a:
+                size_a = size_a[1:]
+            if b == 1 or expand_b:
+                size_b = size_b[1:]
+
+            if well_conditioned:
+                PLU = torch.lu_unpack(*torch.lu(torch.randn(*size_a)))
+                if uni:
+                    # A = L from PLU
+                    A = PLU[1].transpose(-2, -1).contiguous()
+                else:
+                    # A = U from PLU
+                    A = PLU[2].contiguous()
+                    d = A.diagonal(0, -2, -1)
+                    d[d.abs() < 1e-6] = 1.
+            else:
+                A = make_arg(size_a)
+                diag = A.diagonal(0, -2, -1)
+                if uni:
+                    diag.fill_(1.)
+                else:
+                    diag[diag.abs() < 1e-6] = 1.
+                A.triu_()
```

Comment:
What is the fundamental difference between the `well_conditioned` or not so branches?

Reply:
It turns out that if you generate a matrix with `make_arg`, take its `tril_()` part and fill the diagonal with ones to form an upper unitriangular matrix, with high probability it seems to have a very small number as its singular value for matrices of size 60x60 or so. No idea why. This is particularly interesting, as this is a matrix with determinant 1 but very badly conditioned.

For these matrices (the tests in `test_linalg_solve_triangular_large`) I generate these matrices via choosing the L from a PLU decomposition of a Gaussian matrices. This matrix had reasonable eigenvalues, so all good here.

Reply:
There is a utility method that allows you to generate well-conditioned matrices with Gaussian singular values of specific mean/variance which is in `torch/testing/_internal/common_utils` and is `random_well_conditioned_matrix`.

Reply:
In this case I wanted specifically upper/lower triangular matrices, and this was simple enough to be implemented here, as it was just a couple lines.